SSA/Projects/InstCombine/LLVM/Semantics.lean

```diff
@@ -93,13 +128,21 @@ structure NoWrapFlags where
 def add {w : Nat} (x y : IntW w) (flags : NoWrapFlags := {nsw := false , nuw := false}) : IntW w := do
   let x' ← x
   let y' ← y
-  if flags.nsw ∧ x'.msb = y'.msb ∧ (x' + y').msb ≠ x'.msb then
+  if flags.nsw ∧ sadd_overflow x' y' then
     none
-  else if flags.nuw ∧ (x' + y' < x' ∨ x' + y' < y') then
+  else if flags.nuw ∧ uadd_overflow x' y' then
     none
   else
     add? x' y'
 
+  def sadd_overflow_eq {w : Nat} (x y : BitVec w) :
+      (x.msb = y.msb ∧ (x + y).msb ≠ x.msb) = (sadd_overflow x y) := by
+    sorry
```

Comment:
Can you set w to 32 and see if bv_decide proves it?

Reply:
We do not necessarily need a symbolic proof.

Reply:
Try `simp[sadd_overflow]` first so that `bv_decide` can see the defn. As it says:

>  It abstracted the following unsupported expressions as opaque variables: [`BitVec.ofBool (sadd_overflow x y)`]

Therefore, it doesn't know how to interpret the meaning of `sadd_overflow`.

Reply:
This won't help. The definitions are in function of `toNat` or `toInt`. We need a proof to sth that BVDecide can reason about.

Reply:
OK, then we need to write a proof :)

Reply:
Luisa posted the relevant statements on zulip. I hope the proof will follow soon.